docs/src/proposals/bidirectional_quic_reply.md

```diff
@@ -0,0 +1,118 @@
+bidirectional channel once and keep it alive to receive notifications about the transactions that it
+has sent to the cluster. If the connection will be kept open by the server for a predetermined N
+seconds and then will be closed. We can emit these notification at the most important stages of banking
+stage and where transactions are probably lost.
+
+A TPU client which connects to FANOUT_SLOTS/NUMBER_OF_SLOTS_PER_LEADER number of the next leader
+then will initialize a bidirectional channel once with each of the leaders and send the transactions.
+Each leader will then reply how it responded to the transactions and we can easily determine
+the lifecycle of the transaction. We add a mechanism in the TPU client called a bidirectional
+reply handler which will take care of the notifications received by each leader, and forward the
+notifications to a crossbeam channel to be processed by the user of the TPU client.
+
+To avoid loops and over-notifications, each bidirectional channel will be limited to a single validator,
+the validator will not create bidirectional channels with other validators to get notifications.
+The size of the reply message will be a fixed length of N bytes so that the client will exactly
+know that it has received a notification from the server after receiving N bytes.
+Here is the proposed structure:
+
+```rust
+pub const QUIC_REPLY_MESSAGE_SIZE: usize = 256;
+#[derive(Clone, Serialize, Deserialize)]
+pub enum QuicReplyMessage {
+    TransactionMessage {
+        sender_identity: Pubkey,
+        transaction_signature: Signature,
+        message: Vec<u8>, // Is resized to 128 chars in constructor
+        approximate_slot: Slot,
+        padding: [u8; ] // rest
+    },
+}
+```
+
+In the above structure we have given an example for a QuicReplyMessage for a transaction, the message
+length should be exactly 128 chars. A serialized message for type QuicReplyMessage
+should have exact length of QUIC_REPLY_MESSAGE_SIZE which is set to be 256 bytes at this time. If the
+message is larger than 128 characters then it will be truncated and if it is smaller then the rest of characters
+will be set to `0`.
+
+We can additionally add limits so that this feature could not be used as an attack vector.
+
+1. Number of connections in bidirectional stream 5 unstaked and 5 staked.
+2. Timeout for connection on server side : 30 seconds (i.e server will close connection in 30
+seconds)
+3. Additionally server will close the connection when client closes the connection.
+4. Timeout for notification message on client side : 10s
+5. Max number of transaction per connection : 10000
+6. Number of messages in the queue to be sent limited to : 10000
+7. Only staked identities can be used to create a bidirectional connections.
+8. The minimum stake required is 10000 SOLs.
+
+**This addition is over the unidirectional stream connection, so that it will support older clients too.**
```

Comment:
wdyt about creating a separate stream for the back channel rather than building on the existing unidirectional stream connection.  This is strictly a debug feature that's disabled by default and probably rarely recommended on mainnet. Keeping the code paths fully separate from the existing channel seems the best from a security posture perspective, and an easier incremental change to contemplate than one that directly modifies such a mission-critical validator function

Reply:
Ok I understand the hesitation to make these changes in QUIC stage. How about creating a new service that can retrieve notification messages from all the services (banking, replay, etc.). Clients will connect to this service using a bidirectional channel and then send messages about which type of notification they want to subscribe to. We can have all types of protection on it; identity whitelist files, number of subscriptions allowed etc. What do you think?

Reply:
Yeah that approach seems much nicer to me, thanks.
The RpcPubSub service essentially takes this general approach already (but tbc, I'm *not* suggesting we use jsonrpc-pubsub in this application, still bidirectional quic)

Reply:
Yeah I understand, Ok i will rewrite the proposal reflecting that this is another service of itself. Disabled by default, can be enabled by cli args on the specified ports.

Reply:
Updated.